packages/graph/hash_graph/lib/graph/src/api/rest/middleware.rs

```diff
@@ -0,0 +1,73 @@
+use axum::{
+    body::{Body, Bytes, HttpBody},
+    http::{Request, StatusCode},
+    middleware::Next,
+    response::{IntoResponse, Response},
+};
+use tracing::{enabled, Level};
+
+// *Heavily* inspired by
+// https://github.com/tokio-rs/axum/blob/main/examples/print-request-response/src/main.rs
+
+/// A development-environment-focused Axum Handler function to buffer the body of requests and
+/// responses and log them. Overhead should be minimal when not in `Level::Trace`
+pub(super) async fn log_request_and_response(
+    request: Request<Body>,
+    next: Next<Body>,
+) -> Result<impl IntoResponse, (StatusCode, String)> {
+    let mut request = request;
+    if enabled!(Level::TRACE) {
```

Comment:
The docs imply this is meant to be used in `Subscriber` impls but I have checked on the Discord and it should work anywhere. I have also checked it correctly disables this code if not in trace logging.